tcms/core/tests/test_ajax.py

```diff
@@ -185,3 +182,34 @@ def test_remove_tag_from_test_run(self):
 
         self.assertEqual(response.status_code, http.client.OK)
         self.assertEqual(self.test_run.tag.count(), 0)
+
+
+class Test_Tag_Render(Test_Tag_Test):
+
+    def test_render_plan(self):
+        self.test_plan.add_tag(self.test_tag)
+
+        for _ in range(0, 3):
+            TestPlanFactory().add_tag(self.test_tag)
+
+        response = self.client.get(self.url, {
+            'plan': self.test_plan.plan_id
+        })
+
+        self.assertEqual(response.status_code, http.client.OK)
+        self.assertEqual(self.test_plan.tag.count(), 1)
+        self.assertContains(response, '>4</a>')
+
+    def test_render_case(self):
```

Comment:
Move the setup portion of both tests into a setup method. Also add tags to TestRun(s) because rendering also counts runs. Also make sure you are using at least 2 different tags, with different count of objects they are attached to.

Then you can split the assertion operations into a helper method (something like assert_tags or similar). The helper method will assert on all numbers together, not in isolation.

Then in your `test_*` methods simply fetch the response and pass it to the helper assert method!

This will make the current test much more robust and less likely to miss something.